docs/source-pytorch/fabric/advanced/collectives.rst

```diff
@@ -4,45 +4,237 @@
+
+.. figure:: https://pl-flash-data.s3.amazonaws.com/fabric/docs/collectives/ranks.jpeg
+   :alt: The different type of process ranks: Local, global, node.
+   :width: 100%
+
+Here is how you launch multiple processes in Fabric:
 
 .. code-block:: python
 
-    fabric = Fabric()
+    from lightning.fabric import Fabric
 
-    # Transfer and concatenate tensors across processes
-    fabric.all_gather(...)
+    # Devices and num_nodes determine how many processes there are
+    fabric = Fabric(devices=2, num_nodes=3)
+    fabric.launch()
 
-    # Transfer an object from one process to all the others
-    fabric.broadcast(..., src=...)
+Learn more about :doc:`launching distributed training <../fundamentals/launch>`.
+And here is how you access all rank and world size information:
 
-    # The total number of processes running across all devices and nodes.
-    fabric.world_size
+.. code-block:: python
 
-    # The global index of the current process across all devices and nodes.
+    # The total number of processes running across all devices and nodes
+    fabric.world_size  # 2 * 3 = 6
+
+    # The global index of the current process across all devices and nodes
     fabric.global_rank
 
-    # The index of the current process among the processes running on the local node.
+    # The index of the current process among the processes running on the local node
     fabric.local_rank
 
-    # The index of the current node.
+    # The index of the current node
     fabric.node_rank
 
-    # Whether this global rank is rank zero.
-    if fabric.is_global_zero:
-        # do something on rank 0
+    # Do something only on rank 0
+    if fabric.global_rank == 0:
         ...
 
-    # Wait for all processes to enter this call.
+
+.. _race conditions:
+
+Avoid race conditions
+=====================
+
+Access to the rank information helps you avoid *racing conditions* which could crash your script or lead to corrupted data.
+Such conditions can occur when multiple processes are trying to write to the same file all at the same time, for example, in the case of writing a checkpoint file or downloading a dataset.
+Avoid this from happening by guarding your logic with a rank check:
+
+.. code-block:: python
+
+    # Only write files from one process (rank 0) ...
+    if fabric.global_rank == 0:
+        with open("output.txt", "w") as file:
+            file.write(...)
+
+    # ... or save from all processes but don't write to the same file
+    with open(f"output-{fabric.global_rank}.txt", "w") as file:
+        file.write(...)
+
+    # Multi-node: download a dataset, the filesystem between nodes is shared
+    if fabric.global_rank == 0:
+        download_dataset()
+
+    # Multi-node: download a dataset, the filesystem between nodes is NOT shared
+    if fabric.local_rank == 0:
+        download_dataset()
+
+----
+
+
+*******
+Barrier
```

Comment:
Can you put these collective operations into it's own section called collectives?

And link PyTorch's docs: https://pytorch.org/docs/stable/distributed.html

This way this is ready for when our collectives are integrated.

Reply:
I tried it, but it doesn't look so good because the titles get really small if they are grouped under yet another level. Actually, the whole page was already meant to be the section where we list the collectives.